core/src/main/java/io/questdb/cairo/mv/MatViewRefreshJob.java

```diff
@@ -234,35 +290,50 @@ private boolean insertAsSelect(
+                    long commitTarget = batchSize;
+                    rowCount = 0;
+
+                    intervalIterator.toTop(intervalStep);
+                    while (intervalIterator.next()) {
+                        refreshExecutionContext.setRange(intervalIterator.getTimestampLo(), intervalIterator.getTimestampHi());
+                        try (RecordCursor cursor = factory.getCursor(refreshExecutionContext)) {
+                            final Record record = cursor.getRecord();
+                            while (cursor.hasNext()) {
+                                TableWriter.Row row = tableWriter.newRow(record.getTimestamp(cursorTimestampIndex));
+                                copier.copy(record, row);
+                                row.append();
+                                if (++rowCount >= commitTarget) {
+                                    tableWriter.commit();
+                                    commitTarget = rowCount + batchSize;
+                                }
                             }
                         }
                     }
                     break;
                 } catch (TableReferenceOutOfDateException e) {
                     factory = Misc.free(factory);
-                    if (i == maxRecompileAttempts - 1) {
-                        LOG.info().$("base table is under heavy DDL changes, will reattempt refresh later [view=").$(viewDef.getMatViewToken())
-                                .$(", recompileAttempts=").$(maxRecompileAttempts)
+                    if (i == maxRetries) {
+                        LOG.info().$("base table is under heavy DDL changes, will retry refresh later [view=").$(viewDef.getMatViewToken())
+                                .$(", recompileAttempts=").$(maxRetries)
                                 .I$();
                         viewGraph.enqueueIncrementalRefresh(viewDef.getMatViewToken());
                         return false;
                     }
                 } catch (Throwable th) {
                     factory = Misc.free(factory);
-                    refreshFailState(state, refreshTimestamp, th.getMessage());
-                    throw th;
+                    if (CairoException.isCairoOomError(th) && i < maxRetries && intervalStep > 1) {
+                        intervalStep /= 2;
+                        LOG.info().$("query failed with out-of-memory, retrying with a reduced intervalStep")
+                                .$(" [view=").$(viewDef.getMatViewToken())
+                                .$(", intervalStep=").$(intervalStep)
+                                .$(", error=").$(((CairoException) th).getFlyweightMessage())
+                                .I$();
+                    } else {
+                        refreshFailState(state, refreshTimestamp, th.getMessage());
+                        throw th;
+                    }
                 }
+                Os.sleep(200);
```

Comment:
We should sleep only after an OOM exception.

nit: also, the interval should be configurable.

Reply:
My thinking is this: if there's heavy DDL load on a table, causing recompilation, that load should be time-bounded. Retrying 10 times as fast as you can seems a worse option than retrying five times a second over two seconds.

Reply:
DDL load is usually done in bursts, so it's ok to spin. That's why we have no sleep for SQL recompilation in, say, INSERT AS SELECT.